### ISSUE TITLE
Evaluate Swift 5 performance

### ISSUE THREAD

**Issue by djones6** [IBM-Swift/Kitura#1372]
## Context and Description
- Swift 5 vs Swift 4.2
- Swift 5 before + after String utf8 changes
  - Identify places where we convert between bytes and Strings, and evaluate whether we are getting the benefits of the rework
- Flame graph analysis to investigate result of above, or spot any new areas of interest

**Comment by djones6** [comment]
This is blocked by #1371

**Comment by djones6** [comment]
During development of the fix for #1374 (IBM-Swift/Kitura-net#284), I ran a brief performance comparison of Kitura hello-world built on Swift 4.2.1 versus a late December Swift 5 snapshot:
```
               | Throughput (req/s)      | CPU (%) | Mem (kb)     | Latency (ms)                   | good
Implementation | Average    | Max        | Average | Avg peak RSS | Average  | 99%      | Max      | iters
---------------|------------|------------|---------|--------------|----------|----------|----------|-------
       base421 |    63448.3 |    66149.0 |    97.4 |        26174 |      2.0 |      4.0 |     21.2 |    10
        gcd421 |    63671.2 |    65527.9 |    97.2 |        26165 |      2.0 |      3.8 |    202.0 |    10
		
Swift 5:
           gcd |    53529.5 |    54392.8 |    99.1 |        29218 |      2.3 |      3.0 |    202.0 |    10
```
In the results above, `base421` is Kitura prior to the fix for #1374, `gcd421` is the fix that was delivered, and `gcd` is the same code built with Swift 5.  These numbers were collected on Ubuntu 16.04.

This is a significant (16%) regression between 4.2.1 and 5.0.  An initial flame graph comparison implies that we are spending significantly more time in `(extension in Foundation):Swift.StringProtocol.lengthOfBytes(using: (extension in Foundation):Swift.String.Encoding) -> Swift.Int`, which is called from `KituraNet.HTTPServerResponse.writeToSocketThroughBuffer()`.

This is quite interesting as Kitura was recently patched to use an alternative for just that function: https://github.com/IBM-Swift/Kitura/commit/5542795a8f2b833f0b3529e8603f3a15e006fee8

**Comment by djones6** [comment]
Applying the same fix to Kitura-net resolves this (Swift 5.0 comes out marginally ahead):
```
               | Throughput (req/s)      | CPU (%) | Mem (kb)     | Latency (ms)                   | good
Implementation | Average    | Max        | Average | Avg peak RSS | Average  | 99%      | Max      | iters
---------------|------------|------------|---------|--------------|----------|----------|----------|-------
           421 |    63587.4 |    65449.0 |    97.2 |        26145 |      2.0 |      5.4 |    202.8 |    10
       421_fix |    64299.5 |    66892.2 |    97.5 |        26094 |      2.0 |      4.1 |    203.0 |    10
            50 |    52964.2 |    54838.8 |    99.0 |        29153 |      2.4 |      3.4 |    202.0 |    10
        50_fix |    67132.1 |    68740.2 |    97.6 |        29196 |      1.9 |      4.4 |    201.5 |    10
```
I'll raise a PR for this.

**Comment by ianpartridge** [comment]
We've achieved parity.  We'll use a new issue for further performance opportunities.
